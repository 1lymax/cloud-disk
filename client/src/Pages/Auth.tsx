import {useSnackbar} from 'notistack';
import {LocalStorage} from "ts-localstorage";
import React, {useEffect, useState} from 'react';
import LoadingButton from '@mui/lab/LoadingButton';
import HowToRegIcon from '@mui/icons-material/HowToReg';
import LockOutlinedIcon from '@mui/icons-material/LockOutlined';
import {NavLink, useLocation, useNavigate} from "react-router-dom";
import {Avatar, Box, Grid, Link, TextField, Typography} from '@mui/material';

import {useAppDispatch, useAppSelector} from "../hooks/hooks";
import {userAPI} from "../services/UserService";
import {getErrorMessage} from "../utils/getErrorMessage";
import {setAuth, setCurrentUser} from "../store/reducers/UserSlice";
import {ACCESS_TOKEN, DISK_ROUTE, LOGIN_ROUTE, REGISTRATION_ROUTE} from "../utils/consts";


const Auth = () => {
    const location = useLocation()
    const navigate = useNavigate()
    const dispatch = useAppDispatch()
    const {enqueueSnackbar} = useSnackbar()
    const isAuth = useAppSelector(state => state.userState.isAuth)
    const isLogin = location.pathname === LOGIN_ROUTE
    const [email, setEmail] = useState<string>('');
    const [password, setPassword] = useState<string>('');
    const [login, {
        isLoading: loginLoading,
        error: loginError,
        isSuccess: loginSuccess,
        data: loginData
    }] = userAPI.useLoginMutation()
    const [registration, {
        isLoading: regLoading,
        error: regError,
        isSuccess: regSuccess,
    }] = userAPI.useRegistrationMutation()

    const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
        event.preventDefault();
        if (isLogin)
            login({email, password})
        else
            registration({email, password})
    };

    useEffect(() => {
        if (loginError || regError)
            enqueueSnackbar(getErrorMessage(loginError || regError), {variant: "error"});
    }, [loginError, regError]);

    useEffect(() => {
        if (loginSuccess) {
            console.log('login success')
            enqueueSnackbar("Successfully logged in", {variant: "success"});
            dispatch(setCurrentUser(loginData?.user));
            dispatch(setAuth(true))
            LocalStorage.setItem(ACCESS_TOKEN, loginData?.token);
        }
    }, [loginSuccess]);

    useEffect(() => {
        console.log('navigate to disk')
        if (isAuth) navigate(DISK_ROUTE)
    }, [isAuth]);

    useEffect(() => {
        if (regSuccess) {
            enqueueSnackbar("User was created.", {variant: "success"});
            login({email, password})
        }
    }, [regSuccess]);


    return (
        <Box
            sx={{
                marginTop: 8,
                display: 'flex',
                flexDirection: 'column',
                alignItems: 'center',
            }}
        >
            <Avatar sx={{m: 1, bgcolor: 'secondary.main'}}>
                {isLogin ? <LockOutlinedIcon/> : <HowToRegIcon/> }
            </Avatar>
            <Typography component="h1" variant="h5">
                {isLogin ? 'Sign in' : 'Sign up'}
            </Typography>
            <Box component="form" onSubmit={handleSubmit} noValidate sx={{mt: 1}}>
                <TextField
                    id="email"
                    required
                    fullWidth
                    autoFocus
                    name="email"
                    margin="normal"
                    value={email}
                    autoComplete="email"
                    label="Email Address"
                    onChange={e => setEmail(e.target.value)}
                />
                <TextField
                    required
                    fullWidth
                    id="password"
                    margin="normal"
                    type="password"
                    name="password"
                    label="Password"
                    value={password}
                    autoComplete="current-password"
                    onChange={e => setPassword(e.target.value)}
                />
                {/*<FormControlLabel*/}
                {/*    control={<Checkbox value="remember" color="primary"/>}*/}
                {/*    label="Remember me"*/}
                {/*/>*/}
                <LoadingButton
                    fullWidth
                    type="submit"
                    loading={loginLoading || regLoading}
                    variant="contained"
                    sx={{mt: 3, mb: 2}}
                >
                    {isLogin ? 'Sign in' : 'Sign up'}
                </LoadingButton>
                <Grid container>
                    <Grid item xs>
                        <Link href="#" variant="body2">
                            Forgot password?
                        </Link>
                    </Grid>
                    <Grid item>
                        <NavLink to={isLogin ? REGISTRATION_ROUTE : LOGIN_ROUTE}>
                            <Typography variant="body2">
                                {isLogin ? 'Don\'t have an account? Sign up' : 'Already registered? Sign in'}
                            </Typography>
                        </NavLink>
                    </Grid>
                </Grid>
            </Box>
        </Box>
    );
};

export default Auth;