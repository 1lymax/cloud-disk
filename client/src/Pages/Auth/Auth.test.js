import userEvent from "@testing-library/user-event";
import {screen, waitFor} from '@testing-library/react';
import Auth from "./Auth";
import {renderTestApp} from "../../tests/helpers/renderTestApp";



describe('Auth', () => {
	test('Page snapshot', async () => {
		renderTestApp(null, {route: '/login'});
		const authPage = screen.getByTestId('auth-page')
		await expect(authPage).toMatchSnapshot()

		await userEvent.type(screen.getByTestId('login-input'), 'temp@gmail.com')
		await userEvent.type(screen.getByTestId('password-input'), 'qwerty')
		await userEvent.click(screen.getByTestId('login-btn'))

		//expect redirecting to disk page
		await waitFor(() => {
			// snackbar message
			expect(screen.getByText('Successfully logged in')).toBeInTheDocument()
			expect(screen.getByTestId('disk-page')).toBeInTheDocument()
		})

	});
});
