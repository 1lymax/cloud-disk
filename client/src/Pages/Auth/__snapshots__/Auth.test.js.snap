// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Page snapshot 1`] = `
<div
  class="MuiBox-root css-binzgt"
  data-testid="auth-page"
>
  <div
    class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1999axt-MuiAvatar-root"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="LockOutlinedIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
      />
    </svg>
  </div>
  <h1
    class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
  >
    Sign in
  </h1>
  <form
    class="MuiBox-root css-164r41r"
    novalidate=""
  >
    <div
      class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-focused Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="email"
        id="email-label"
      >
        Email Address
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          autocomplete="email"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          data-testid="login-input"
          id="email"
          name="email"
          required=""
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Email Address
               
              *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="password"
        id="password-label"
      >
        Password
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          autocomplete="current-password"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          data-testid="password-input"
          id="password"
          name="password"
          required=""
          type="password"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Password
               
              *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1qudo6r-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
      data-testid="login-btn"
      id=":r2:"
      tabindex="0"
      type="submit"
    >
      Sign in
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <div
      class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
      >
        <a
          class="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineAlways css-101ca9i-MuiTypography-root-MuiLink-root"
          href="src/Pages/Auth/Auth#"
        >
          Forgot password?
        </a>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <a
          class=""
          data-testid="registration-link"
          href="/registration"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            Don't have an account? Sign up
          </p>
        </a>
      </div>
    </div>
  </form>
</div>
`;
